numpy/_core/src/multiarray/iterators.c

```diff
@@ -561,195 +489,157 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
 NPY_NO_EXPORT PyObject *
 iter_subscript(PyArrayIterObject *self, PyObject *ind)
 {
-    PyArray_Descr *indtype = NULL;
-    PyArray_Descr *dtype;
-    npy_intp start, step_size;
-    npy_intp n_steps;
-    PyArrayObject *ret;
-    char *dptr;
-    int size;
-    PyObject *obj = NULL;
-    PyObject *new;
+    PyObject *ret = NULL;
+
+    int index_type;
+    int index_num = -1;
+    int ndim, fancy_ndim;
+    npy_intp start, stop, step, n_steps;
+    npy_index_info indices[NPY_MAXDIMS * 2 + 1];
+
+    PyArray_Descr *dtype = PyArray_DESCR(self->ao);
+    npy_intp dtype_size = dtype->elsize;
     NPY_cast_info cast_info = {.func = NULL};
 
-    if (ind == Py_Ellipsis) {
-        ind = PySlice_New(NULL, NULL, NULL);
-        obj = iter_subscript(self, ind);
-        Py_DECREF(ind);
-        return obj;
-    }
-    if (PyTuple_Check(ind)) {
-        int len;
-        len = PyTuple_GET_SIZE(ind);
-        if (len > 1) {
-            goto fail;
-        }
-        if (len == 0) {
-            Py_INCREF(self->ao);
-            return (PyObject *)self->ao;
-        }
-        ind = PyTuple_GET_ITEM(ind, 0);
+    /* Prepare the indices */
+    index_type = prepare_index_noarray(1, &self->size, ind, indices, &index_num,
+        &ndim, &fancy_ndim, 1, 1);
+
+    if (index_type < 0) {
+        return NULL;
     }
 
-    /*
-     * Tuples >1d not accepted --- i.e. no newaxis
-     * Could implement this with adjusted strides and dimensions in iterator
-     * Check for Boolean -- this is first because Bool is a subclass of Int
-     */
-    PyArray_ITER_RESET(self);
+    else if (indices[0].type == HAS_NEWAXIS) {
+        PyErr_SetString(PyExc_IndexError,
+            "only integers, slices (`:`), ellipsis (`...`) and integer or boolean "
+            "arrays are valid indices"
+        );
+        goto finish;
+    }
 
-    if (PyBool_Check(ind)) {
-        int istrue = PyObject_IsTrue(ind);
-        if (istrue == -1) {
-            goto fail;
+    // Single ellipsis index
+    else if (index_type == HAS_ELLIPSIS) {
+        ind = PySlice_New(NULL, NULL, NULL);
+        if (ind == NULL) {
+            goto finish;
         }
-        if (istrue) {
-            return PyArray_ToScalar(self->dataptr, self->ao);
+
+        ret = iter_subscript(self, ind);
+        Py_DECREF(ind);
+        goto finish;
+    }
+
+    // Single boolean index
+    else if (indices[0].type == HAS_0D_BOOL) {
+        DEPRECATE("Indexing flat iterators with a 0-dimensional boolean index is "
```

Comment:
Needs:
* A comment `/* Deprecated 2025-..., NumPy 2.4 */`
* an error check
* And a test in `test_deprecations()` which would have failed the missing error check.
* Please end the message with `(Deprecated NumPy 2.4)` or similar, to give people a sense of how recent this happened.